src/messaging/messaging-service.js

```diff
@@ -0,0 +1,129 @@
+export default class MessagingService {
+  constructor (destinations = []) {
+    /**
+     * A map object where outgoing messages will be stored. The key is the message ID and the value is an object
+     * that stores details about the message being sent.
+     *
+     * @type {Map<string, StoredRequest>}
+     */
+    this._messages = new Map()
+
+    /**
+     * A map object where outgoing messages will be stored. The key is a destination name and the value is
+     * the Destination object.
+     *
+     * @type {Map<string, Destination>}
+     */
+    this._destinations = new Map()
+
+    // If provided as a singular value convert destination into an array
+    if (!Array.isArray(destinations)) { destinations = [destinations] }
+    destinations.forEach(destination => this.registerDestination(destination))
+  }
+
+  /**
+   * Registers destination by adding it to the destinations map and setting a response callback.
+   *
+   * @param {Destination} destination - A destination object to register.
+   */
+  registerDestination (destination) {
+    this._destinations.set(destination.name, destination)
+    destination.registerResponseCallback(this.dispatchMessage.bind(this))
+  }
+
+  /**
+   * A function to handle incoming messages.
+   *
+   * @param {ResponseMessage} message - An incoming response message.
+   */
+  dispatchMessage (message) {
+    if (!ResponseMessage.isResponse(message)) {
+      console.error('A message not following a response format will be ignored:', message)
+      return
+    }
+
+    if (!this._messages.has(message.requestHeader.ID)) {
+      console.error(`Ignoring a message with request ID ${message.requestHeader.ID} not registered in a request list`, message)
+      return
+    }
+    const requestInfo = this._messages.get(message.requestHeader.ID)
+    window.clearTimeout(requestInfo.timeoutID) // Clear a timeout
+    const responseCode = message.responseCode
+
+    if (responseCode === ResponseMessage.responseCodes.ERROR) {
+      // There was an error returned. An error info is in the message body.
+      requestInfo.reject(message.body)
+    } else {
+      // Request was processed without errors
+      requestInfo.resolve(message)
+    }
+    this._messages.delete(message.requestHeader.ID) // Remove request info from the map
+  }
+
+  /**
+   * Registers an outgoing request within a request map. Returns a promise that will be fulfilled when
+   * a response will be received or rejected when a timeout will expire.
+   *
+   * @param {RequestMessage} request - An outgoing request.
+   * @param {number} timeout - A number of milliseconds we'll wait for response before rejecting a promise.
+   * @returns {Promise} - A promise that will be resolved with the message response or rejected with an error info.
+   */
+  registerRequest (request, timeout = 10000) {
+    let storedRequest = new StoredRequest(request) // eslint-disable-line prefer-const
+    this._messages.set(request.ID, storedRequest)
```

Comment:
as with destinations, should we check to be sure that we aren't registering a message with a duplicate ID?

Reply:
We're generating all IDs as RFC4122 v4 UUID: https://github.com/kelektiv/node-uuid#version-4. The chance of collision is very minimal, I believe. So I decided to spare a few computing cycles and skip the check. What do you think? Would it be safe enough for us?

Reply:
The only thing is that the ensuring of the uniqueness of the request.id is not the responsibility of the messaging service, so there is no guarantee in this code, which relies on the ids being unique to function correctly, that it is getting unique ids. I'd prefer to see the check.

Reply:
This is a good point. I've added a check for a request ID to be unique.